src/OrbitGl/QtTextRenderer.cpp

```diff
@@ -0,0 +1,228 @@
+  }
+  // The entire string does not fit. We try to fit something of the form
+  // leading_text + "... " + trailing_text
+  // where leading_text is a variable amount of characters from the beginning of the string and
+  // trailing_text is specified by the trailing_chars_length parameter.
+  const std::string text_as_string(text);
+  const std::string trailing_text = text_as_string.substr(text_length - trailing_chars_length);
+  const std::string leading_text = text_as_string.substr(0, text_length - trailing_chars_length);
+  const size_t leading_length = leading_text.length();
+  const std::string ellipsis_plus_trailing_text = std::string("... ") + trailing_text;
+  size_t fitting_chars_count = 0;
+  std::string candidate_string =
+      leading_text.substr(0, fitting_chars_count).append(ellipsis_plus_trailing_text);
+
+  // Test if we can fit the minimal ellipsised string.
+  if (GetStringWidth(candidate_string.c_str(), formatting.font_size) > max_width) {
+    // We can't fit any ellipsised string: Even with no leading character the thing becomes too
+    // long. So we let AddText truncate the entire string.
+    Vec2 dims;
+    AddText(text, x, y, z, formatting, nullptr, &dims);
+    return dims[0];
+  }
+
+  // Binary search between 0 and leading_length (we know 0 works and leading_length does not).
+  size_t idx_min = 0;
+  size_t idx_max = leading_length;
+  while (idx_max - idx_min > 1) {
+    size_t candidate_idx = (idx_min + idx_max) / 2;
+    candidate_string = leading_text.substr(0, candidate_idx).append(ellipsis_plus_trailing_text);
+    if (GetStringWidth(candidate_string.c_str(), formatting.font_size) > max_width) {
+      idx_max = candidate_idx;
+    } else {
+      idx_min = candidate_idx;
+    }
+  }
+  fitting_chars_count = idx_min;
+  candidate_string =
+      leading_text.substr(0, fitting_chars_count).append(ellipsis_plus_trailing_text);
+  AddText(candidate_string.c_str(), x, y, z, formatting);
+  return GetStringWidth(candidate_string.c_str(), formatting.font_size);
+}
+
+float QtTextRenderer::GetStringWidth(const char* text, uint32_t font_size) {
+  const std::vector<std::string> lines = absl::StrSplit(std::string(text), '\n');
+  float max_width = 0.f;
+  QFont font = QFontDatabase::systemFont(QFontDatabase::GeneralFont);
+  font.setPixelSize(static_cast<int>(font_size));
+  QFontMetrics metrics(font);
+  for (const std::string& line : lines) {
+    max_width = std::max(
+        max_width, viewport_->ScreenToWorld(Vec2i(metrics.horizontalAdvance(line.c_str()), 0))[0]);
+  }
+  return max_width;
+}
+
+float QtTextRenderer::GetStringHeight(const char* text, uint32_t font_size) {
+  const float number_of_lines = static_cast<float>(std::count(text, text + strlen(text), '\n') + 1);
+  QFont font = QFontDatabase::systemFont(QFontDatabase::GeneralFont);
+  font.setPixelSize(static_cast<int>(font_size));
+  QFontMetrics metrics(font);
+  return number_of_lines * viewport_->ScreenToWorld(Vec2i(0, metrics.height()))[1];
```

Comment:
Have you seen that `GetStringHeight` was previously returning the height of just one line? I'm sure that your implementation is the correct one, but did you check if we broke some multi-line text?